File: static/src/data/mapper_000.ts
import { Mapper } from "./memory_mapper";

export class Mapper000 extends Mapper {

    private _nPrgBanks: number;
    private _nChrBanks: number;

    constructor(nPrgBanks: number, nChrBanks: number) {
        super();
        this._nPrgBanks = nPrgBanks;
        this._nChrBanks = nChrBanks;
    }

    public MapRead(addr: number): number {
        let mapped = -1;

        if(addr >= 0x0000 && addr <= 0x1FFF) {
            mapped = addr;
        }

        if(addr >= 0x8000 && addr <= 0xFFFF) {
            mapped = addr & (this._nPrgBanks > 1 ? 0x7FFF : 0x3FFF);
        }
        return mapped;

    }

    public MapWrite(addr: number): number {
        //add cart RAM option
        let mapped = -1;
        if(addr >= 0x8000 && addr <= 0xFFFF) {
            mapped = addr & (this._nPrgBanks > 1 ? 0x7FFF : 0x3FFF);
        }
        return mapped;
    }

}

File: static/wasm/src/mapper_000.cpp

#include "mapper_000.h"

Mapper000::Mapper000(uint8_t prgBanks, uint8_t chrBanks) : Mapper(prgBanks, chrBanks) {

}

uint16_t Mapper000::MapRead(uint16_t address, uint8_t &data) {
    uint16_t mapped = address;
    if(address >= 0x8000 && address <= 0xFFFF) {
        mapped = address & (_prgBanks > 1 ? 0x7FFF : 0x3FFF);
    }
    return mapped;
}

uint16_t Mapper000::MapWrite(uint16_t address, uint8_t &data) {    
    uint16_t mapped = address;
    return mapped;
}